tsdb/head.go

```diff
@@ -524,23 +531,49 @@ Outer:
 					seriesCreationErr = err
 					break Outer
 				}
-
-				if created {
-					// If this series gets a duplicate record, we don't restore its mmapped chunks,
-					// and instead restore everything from WAL records.
-					series.mmappedChunks = mmappedChunks[series.ref]
+				if !created {
+					// There's already a different ref for this series.
+					multiRef[s.Ref] = series.ref
+				}
+				mmc := mmappedChunks[series.ref]
+				if len(series.mmappedChunks) == 0 && len(mmc) > 0 {
+					series.mmappedChunks = mmc
 
 					h.metrics.chunks.Add(float64(len(series.mmappedChunks)))
 					h.metrics.chunksCreated.Add(float64(len(series.mmappedChunks)))
 
 					if len(series.mmappedChunks) > 0 {
 						h.updateMinMaxTime(series.minTime(), series.maxTime())
 					}
-				} else {
-					// TODO(codesome) Discard old samples and mmapped chunks and use mmap chunks for the new series ID.
+				} else if len(mmc) > 0 && len(series.mmappedChunks) > 0 &&
+					series.mmappedChunks[len(series.mmappedChunks)-1].maxTime > mmc[len(mmc)-1].maxTime {
+					// These new m-mapped chunks in the future for the same series.
+					// Hence replace the old m-mapped chunks with new chunks.
+
+					if overlapsClosedInterval(
+						series.mmappedChunks[0].minTime,
+						series.mmappedChunks[len(series.mmappedChunks)-1].maxTime,
+						mmc[0].minTime,
+						mmc[len(mmc)-1].maxTime,
+					) {
+						// TOOD(codesome) return this instead.
+						panic("overlapping mmapped chunks")
```

Comment:
this? 
I guess todo for this PR

Reply:
Yes, TODO is for this PR. Still WIP.

Reply:
Will fix this in https://github.com/prometheus/prometheus/pull/7438